lib/tasks/bulk.rake

```diff
@@ -11,4 +11,80 @@ namespace :bulk do
       DeliveryRequestWorker.perform_async_in_queue(id, queue: :delivery_immediate)
     end
   end
+
+  desc "Switch immediate subscribers of the specified list slugs to daily digest"
+  task switch_to_daily_digest: :environment do |_t, args|
```

Comment:
I'm just commenting here based on https://github.com/alphagov/email-alert-api/pull/1354.

I feel like this doesn't belong in the `bulk` namespace because in my mind, the "bulk email" is a very specific thing which sends a single email out to multiple subscriber list. The task being added here is a different kind of email which also sends to multiple subscribe lists, but isn't a "bulk email".

I actually think having a `bulk` namespace is a bit confusing and I wonder if restructuring the tasks to be like this would make the distinction clearer (and perhaps avoid needing to do https://github.com/alphagov/email-alert-api/pull/1354):

- `emails:send_bulk`
- `emails:send_switch_to_daily_digest`
- ... any other new emails builders go here

Reply:
@thomasleese it sounds like you're seeing the "bulk" namespace and the "email" task as a single entity. Whereas I'm seeing them as separate things: a namespace to put bulk tasks (we have to put them somewhere, after all), and a task that describes what kind of bulk operation it does.

> https://github.com/alphagov/email-alert-api/pull/1354 is making a similar distinction: a generic name for specific code makes it harder to understand its purpose.

The "email" task could do with a slightly longer name to describe what it does, but for now it's not in competition with any other bulk email task. I don't think "emails" would be appropriate namespace for this new task, because the email sending bit is only part of its behaviour.

Does that help?

Reply:
Yeah, that makes sense. I think this might just be a personal preference thing.

To me, we have plenty of tasks in Email Alert API which perform bulk operations, but their defining characteristic is not the fact that they are bulk, it's _what_ they actually do: whether that be sending emails or managing subscriptions. For example, [we've got three tasks in the `manage` namespace](https://github.com/alphagov/email-alert-api/blob/master/lib/tasks/manage.rake#L117-L144) which are definitely bulk operations and therefore if we were going to stick with this naming scheme, we should probably move them into the `bulk` namespace.

Perhaps my complaint is actually with the naming of the `manage` namespace (and by extension the `troubleshoot` namespace) in that it's very generic, it doesn't really explain what the tasks do and makes it difficult to place new tasks. Thinking about it further, I agree with you that the task in this PR doesn't just send emails and therefore doesn't belong in an `emails` namespace. I'm actually thinking that my ideal structure would be something like this:

- `emails:send_bulk`
- [`emails:test:send_to_subscriber`](https://github.com/alphagov/email-alert-api/blob/4ba89df82a105ce2d88c93058fa9b6f84cf40ce2/lib/tasks/troubleshoot.rake#L20-L30)
- [`subscribers:change_email_address`](https://github.com/alphagov/email-alert-api/blob/4ba89df82a105ce2d88c93058fa9b6f84cf40ce2/lib/tasks/manage.rake#L79)
- `subscribers:switch_to_digest`

I realise though that this is kind of derailing your PR and beyond the scope. If you're interested I'm happy to talk about it further, but I won't block your PR going in as it is now!

Reply:
Thanks @thomasleese. Good to know this isn't a blocker. I agree that two of those tasks you mentioned are indeed bulk operations. However, "unsubscribe all" reflects an action a user can do themselves - why do we even need that?

Anyway, I think there's still some ambiguity with the namespaces, even with your suggestion. When we're writing and finding rake tasks, I think we're more concerned with the context (e.g. "2ndline" or "data_migration"), rather than the scope ("subscribers" or "bulk"). How about this?

## 2ndline support tasks

- `2ndline:view_subscriptions`
- `2ndline:view_emails`
- ~`manage:change_email_address`~ (users can do this themselves?)
- ~`manage:unsubscribe_single_subscription`~ ([the definite need for this](https://github.com/alphagov/email-alert-api/pull/1226) is unclear, user can do this themselves?)
- ~`manage:unsubscribe_all_subscriptions`~ (users can do this themselves?)
- `2ndline:get_notifications_from_notify_by_email_id`
- ~`troubleshoot:get_notifications_from_notify`~ (not sure why we need this if we have the above)
- `2ndline:delivery_to_test_email`
- ~`troubleshoot:delivery_to_subscriber`~ (not sure why we need this if we have the above)
- `2ndline:resend_failed_emails`
- `2ndline:send_bulk_to_subscriber_list`

## Data migration tasks

- `data_migration:move_all_subscribers`
- `data_migration:find_subscriber_list_by_title` ([precursor to the next one](https://github.com/alphagov/email-alert-api/pull/799))
- `data_migration:update_subscriber_list`
- `data_migration:switch_to_daily_digest[_experiment]`
- ~`manage:unsubscribe_bulk_from_csv`~ (seems redundant, based on the [original PR](https://github.com/alphagov/email-alert-api/pull/511))

Reply:
Sorry for the delay, finally got some free time to reply to this!

I think your proposal for classifying the Rake tasks is interesting, although I'm not 100% convinced. For example when I ran the job to send the bulk travel advice email, I wasn't doing it while on 2nd line. I think we might be trying to second guess the context that people are running these tasks in whereas grouping the tasks by _what_ they do feels more natural to me. For example, I struggle to see a big difference between what we have now with `manage:` and your proposal of `2ndline:` in terms of clarity of where to put new tasks and where to find existing ones.

In terms of the tasks you've identified as ones we can remove, there have certainly been times when I've used the tasks to manually unsubscribe people who are having problems (and just solving the problem for them seemed easier than helping them solve it for themselves), but I agree that ultimately it's something users can do so I'm also unsure about their need.

Reply:
Ah, the merged PR comment saga continues. I kind of like it.

I think we've both got some good points, and are slowly converging on something agreeable.

## Pointless tasks

It would help to cut out the cruft. Could you tick/comment on the ones you agree we can get rid of?

- [ ] manage:change_email_address
- [ ] manage:unsubscribe_single_subscription
- [ ] manage:unsubscribe_all_subscriptions
- [ ] troubleshoot:get_notifications_from_notify
- [ ] troubleshoot:delivery_to_subscriber
- [ ] manage:unsubscribe_bulk_from_csv

## Namespaces

- You didn't say anything about `data_migration`, so I'm going to assume that's OK. It incorporates both the context and the behaviour of the tasks within it, so seems to meet both our expectations.

- `2ndline` does seem slightly better for some of the tasks. For example, `view_emails` isn't really managing anything. Some of these tasks appear to be defined specifically for 2ndline, so their purpose is "support" or "troubleshoot". I don't really mind about the word here, as long as it's more specific than "manage". What do you prefer?

- `2ndline:send_bulk_to_subscriber_list` seems to be an outlier. Perhaps it could be in its own namespace, which is specific to `custom_email` or `send_email`, or something?!

What do you think? Hopefully that gets us a bit closer to a conclusion.

Reply:
It's been a few days so I've found time to reply to this 😆 

I've ticked the tasks I don't think we need. In terms of the naming, I think `support` is a better name than `manage` and I prefer that it's not tied specifically to `2ndline`.

I agree that perhaps a good solution here to to put `send_bulk_to_subscriber_list` in its own namespace, I think either `emails` or `send_email` is good.

Reply:
Cool, thanks for the review.

First part of the change here: https://github.com/alphagov/email-alert-api/pull/1367